functions/go/apply-setter/main.go

```diff
@@ -0,0 +1,97 @@
+package main
+
+import (
+	"fmt"
+	"os"
+
+	"sigs.k8s.io/kustomize/kyaml/fn/framework"
+	kyaml "sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+//nolint
+func main() {
+	resourceList := &framework.ResourceList{}
+	resourceList.FunctionConfig = map[string]interface{}{}
+
+	cmd := framework.Command(resourceList, func() error {
+		s, err := getSetters(resourceList.FunctionConfig)
+		if err != nil {
+			return fmt.Errorf("failed to get data.setters field from function config: %w", err)
+		}
+		_, err = s.Filter(resourceList.Items)
+		if err != nil {
+			return fmt.Errorf("failed to set setters: %w", err)
+		}
+		return nil
+	})
+
+	cmd.Long = usage()
+	if err := cmd.Execute(); err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+}
+
+func usage() string {
+	return `Set values of resources with setter values.
+
+Configured using a ConfigMap with the following keys:
+
+setters.name: Setter name.
+setters.value: Setter value.
+
+These keys are in a list in path 'data.setters'.
+
+Example:
+
+To set a setter 'color: orange' to all resources:
+
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: my-config
+data:
+  setters:
```

Comment:
I had some comments from last round which is not resolved.

We should standardize the convention of field names here. Options:

1. setters.name and setters.value. Why have a struct called `setters`? Is this for future-proofing in case the function config will contain other metadata besides setter values? In that case, isn't `setterValues` a better name.
2. setterName and setterValue

Same discussion applies to other functions such as set-label.

@Shell32-Natsu

Reply:
I think it's ok to remove the `setters` and other similar fields in other functions.

IMO, a function should support 2 kinds of input format. 

The *simple* one looks like:

```yaml
...
data:
  key1: value1
  key2: value2
...
```
This is compatible with the input format from `configMap`. There is no `fieldSpecs` or other additional configs in this case.

The *complete* one looks like:
```yaml
...
data:
- name: key1
  value: value1
- name: key2
  value: value2
  fieldSpecs: # if applicable
...
```

WDYT?

Reply:
We currently supports top level `fieldSpecs`
```yaml
...
data:
  fieldSpecs: # top level fieldSpecs
  labels
    - name: key1
      value: value1
      fieldSpecs: # per label level fieldSpecs
    - name: key2
      value: value2
```
I think that's why it's not very flat. IMO we can continue to use this format to support advanced use cases unless we can provide the information through OpenAPI.

Like Donny said above, we can introduce a concise format to support the simpler use cases that don't need to specify `fieldSpecs`
```
...
data:
  key1: value1
  key2: value2
```

Reply:
Besides brevity vs verbosity in `pipeline`, an important consideration is also type safety. In general, all function configs need to be proper KRM with an OpenAPI schema. This enables tooling (e.g. CloudCode) to help with authoring function configs (easily done when using `pipeline.config` and referring to a separate file). We should avoid having "pseudo" resources that don't have a defined Schema. If we're really dealing with simple key-value, using ConfigMap should be OK if any arbitrary key and value are acceptable.

One question: What's the resource kind for the "advance" config:

```
data:
  fieldSpecs: # top level fieldSpecs
  labels
    - name: key1
      value: value1
      fieldSpecs: # per label level fieldSpecs
    - name: key2
      value: value2
```

I'd avoid using `data` since it resembles ConfigMap, it should really be a custom resource

Reply:
We are using ConfigMap for the "advanced" config now and I just realize that the `data` field in the ConfigMap can only contain simple key-value pairs so our use case is invalid. So we should define our own config CRD for each function.

Reply:
I guess we can do something similar to what we did for set-annotation and set-label.
When using configMap as input configuration, we use string key-value pairs:
```
data:
  color: orange
```

We can do the client-side CRD in a followup PR.